File: qiskit_alt/electronic_structure.py
import os
from ._julia_project import project
project.ensure_init()

Main = project.simple_import("Main")
QuantumOps = project.simple_import("QuantumOps")
QiskitQuantumInfo = project.simple_import("QiskitQuantumInfo")
from .pauli_operators import jlSparsePauliOp

# TODO: This could probably be moved to QiskitQuantumInfo.jl
#Main.include(os.path.join(julia_src_dir, 'electronic_structure.jl'))
Main.include(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'electronic_structure.jl'))

def Geometry(qiskit_geometry):
    """
    Convert a geometry specification that originated in Python in the qiskit-nature format to
    an ElectronicStructure.Geometry object. pyjulia will have translated the python input to
    a Matrix.

    .. code:
    In [1]: geometry = [['O', [0., 0., 0.]],
    ...:             ['H', [0.757, 0.586, 0.]],
    ...:             ['H', [-0.757, 0.586, 0.]]]

    In [2]: Geometry(geometry)
    Out[2]: <PyCall.jlwrap Geometry{Float64}(Atom{Float64}[Atom{Float64}(:O, (0.0, 0.0, 0.0)), Atom{Float64}(:H, (0.757, 0.586, 0.0)), Atom{Float64}(:H, (-0.757, 0.586, 0.0))])>
    ```
    """
    return Main.qiskit_geometry_to_Geometry(qiskit_geometry)


def fermionic_hamiltonian(geometry, basis):
    """
    Given a qiskit-nature molecular geometry specification and basis set, return
    the electronic Hamiltonian as a QuantumOps.FermiSum.

    The integrals are computed by pyscf.
    """
    jlgeometry = Geometry(geometry) # Convert Python geometry spec to ElectronicStructure.Geometry
    fermi_op = Main.fermionic_hamiltonian(jlgeometry, basis)
    return fermi_op


def jordan_wigner(fermi_op):
    """
    Compute the Jordan-Wigner transform of a QuantumOps.FermiSum representing a Fermionic Hamiltonian.
    Return a qiskit.SparsePauliOp.
    """
    pauli_op = QuantumOps.jordan_wigner(fermi_op)
    spop_jl = QiskitQuantumInfo.SparsePauliOp(pauli_op) # Convert to QiskitQuantumInfo.SparsePauliOp
    spop = jlSparsePauliOp(spop_jl)  # Convert to qisit.quantum_info.SparsePauliOp
    return spop

File: src/qiskit_alt/electronic_structure.jl
# ENV["PYCALL_JL_RUNTIME_PYTHON"] = Sys.which("python") or ''

#using PyCall
using ElectronicStructure: Atom, Geometry, MolecularSpec,
    InteractionOperator, MolecularData

# using ElectronicStructurePySCF: PySCF

using QuantumOps: FermiSum

"""
    qiskit_geometry_to_Geometry(geometry::Matrix)

Convert a geometry specification that originated in Python in the qiskit-nature format to
an `ElectronicStructure.Geometry` object. pyjulia will have translated the python input to
a `Matrix`.

# Example from python:
```python
In [1]: geometry = [['O', [0., 0., 0.]],
   ...:             ['H', [0.757, 0.586, 0.]],
   ...:             ['H', [-0.757, 0.586, 0.]]]

In [2]: Main.qiskit_geometry_to_Geometry(geometry)
Out[2]: <PyCall.jlwrap Geometry{Float64}(Atom{Float64}[Atom{Float64}(:O, (0.0, 0.0, 0.0)), Atom{Float64}(:H, (0.757, 0.586, 0.0)), Atom{Float64}(:H, (-0.757, 0.586, 0.0))])>
```
"""
function qiskit_geometry_to_Geometry(geometry::Matrix)
    (num_atoms, num_components_per_atom) = size(geometry)
    if num_components_per_atom != 2
        throw(ArgumentError("Expecting two elements to specify geometry of a single Atom. Got ", num_components_per_atom))
    end
    atoms = []
    for atom_in in eachrow(geometry)
        element_symbol = atom_in[1]
        coords = atom_in[2]
        atom = Atom(Symbol(element_symbol), (coords...,))
        push!(atoms, atom)
    end
    return Geometry(atoms...)
end

function fermionic_hamiltonian(geometry::Geometry, basis)
    # Construct the specification of the electronic structure problem.
    mol_spec = MolecularSpec(geometry=geometry, basis=basis)

    # Do calculations using `PySCF` and populate a `MolecularData` object with results.
    # `mol_pyscf` holds a constant, a rank-two tensor, and a rank-four tensor.
    mol_pyscf = MolecularData(PySCF, mol_spec);
    iop = InteractionOperator(mol_pyscf);
    return FermiSum(iop)
end


# Not useful. Just call jordan_wigner(fermi_op) directly
# function qubit_hamiltonian(fermi_op)
#     pauli_op = jordan_wigner(fermi_op)
#     return pauli_op
# end

# function qubit_hamiltonian(geometry::Geometry, basis)
#     fermi_op = fermionic_hamiltonian(geometry::Geometry, basis)
#     pauli_op = jordan_wigner(fermi_op)
#     return pauli_op
# end
